### PR TITLE
nodejs_18: fix build with clang 16

### PR DESCRIPTION
**Pull request by reckenrode**
## Description of changes

While node v18 successfully built, it was crashing in staging-next #263535 while building packages. This patch reverts the previous attempt at a fix and builds nodejs_18 with clang 15 instead. paperless-ngx.frontend was confirmed to successfully build with this patch applied.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/web/nodejs/v18.nix [hunk 1]
```diff
@@ -1,8 +1,20 @@
-{ callPackage, fetchpatch, openssl, python3, enableNpm ? true }:
+{ callPackage, lib, overrideCC, pkgs, buildPackages, fetchpatch, openssl, python3, enableNpm ? true }:
 
 let
+  # Clang 16+ cannot build Node v14 due to -Wenum-constexpr-conversion errors.
```
Comment (eliasnaur) [inline pkgs/development/web/nodejs/v18.nix:4]: Nit: did you mean Node v18 here, or did I misunderstand something?

Comment (reckenrode) [reply pkgs/development/web/nodejs/v18.nix:4]: Copy-paste mistake. It’s the same solution as nodejs_14 copied into nodejs_18. I can push an update that corrects the comment.

**Comment by reckenrode** [comment]
I should also note that #264091 is an alternative that addresses the clang issue more generally. That PR would be my preferred solution to building with other clangs (rather than per-derivation manual overrides as necessary), but it would be understandable if it’s too much right now.

File: pkgs/development/web/nodejs/v18.nix (hunk 1 above)
Comment (eliasnaur) [reply pkgs/development/web/nodejs/v18.nix]: Thank you. An update wasn't necessary; I'm just an eager onlooker from the peanut gallery.

**Comment by vcunat** [comment]
> -Wenum-constexpr-conversion errors

Isn't normally the easiest approach to just avoid such errors by using a `-Wno-error=enum-constexpr-conversion` flag or similar?

**Comment by reckenrode** [comment]
> > -Wenum-constexpr-conversion errors
> 
> Isn't normally the easiest approach to just avoid such errors by using a `-Wno-error=enum-constexpr-conversion` flag or similar?

It’s going to be made a hard error in a future version of clang (see llvm/llvm-project#59036). Pinning the version to clang 15 avoids possible future breakage.

**Comment by vcunat** [comment]
OK, I think there's been enough time for feedback.

**Pull request merged**